lib/collection/src/shards/replica_set/snapshots.rs

```diff
@@ -91,6 +96,68 @@ impl ShardReplicaSet {
             return Ok(false);
         }
 
+        let segments_path = LocalShard::segments_path(replica_path);
+
+        let mut snapshot_segments = HashSet::new();
+        let mut snapshot_manifests = SegmentManifests::default();
+
+        for segment_entry in segments_path.read_dir()? {
+            let segment_path = segment_entry?.path();
+
+            if !segment_path.is_dir() {
+                log::warn!(
+                    "segment path {} in extracted snapshot {} is not a directory",
+                    segment_path.display(),
+                    replica_path.display(),
+                );
+
+                continue;
+            }
+
+            let segment_id = segment_path
+                .file_name()
+                .and_then(|segment_id| segment_id.to_str())
+                .expect("segment path ends with a valid segment id");
+
+            let segment_added = snapshot_segments.insert(segment_id.to_string());
+            debug_assert!(segment_added);
+
+            let manifest_path = segment_path.join("segment_manifest.json");
+
+            if !manifest_path.exists() {
+                log::debug!("segment {segment_id} does not contain segment_manifest.json");
+                continue;
+            }
+
+            log::debug!("segment {segment_id} contains segment_manifest.json");
+
+            let manifest: SegmentManifest =
+                serde_json::from_reader(io::BufReader::new(fs::File::open(manifest_path)?))?;
+
+            let manifest_added = snapshot_manifests.add(manifest);
+            debug_assert!(manifest_added);
+        }
+
+        debug_assert!(
+            snapshot_manifests.is_empty() || snapshot_segments.len() == snapshot_manifests.len(),
+            "extracted snapshot {} contains segment_manifest.json for some, but not all segments\n\
+             segments: {:?}\n
+             manifests: {:#?}",
+            replica_path.display(),
+            snapshot_segments,
+            snapshot_manifests,
+        );
+
+        for (segment_id, segment_manifest) in snapshot_manifests.iter() {
```

Comment:
```suggestion
        #[cfg(debug_assertions)]
        for (segment_id, segment_manifest) in snapshot_manifests.iter() {
```

Reply:
Sure, I just expected compiler to optimize the whole loop away on release builds, so I did not bother.

Reply:
Yeah you're probably right.

Reply:
Restructured the loop, so now it actually does something in release mode.